src/ctia/stores/es/crud.clj

```diff
@@ -202,24 +213,125 @@ It returns the documents with full hits meta data including the real index in wh
+             (group-by :result)
+             (into {}
+                   (map (fn [[result items]]
+                          {(keyword result) (map :_id items)}))))]
+    (cond-> {}
+      deleted (assoc :deleted deleted)
+      updated (assoc :updated updated)
+      not_found (assoc-in [:errors :not-found] not_found))))
+
+(defn check-and-prepare-bulk
+  [{{{:keys [get-in-config]} :ConfigService} :services
+    :as conn-state}
+   ids
+   ident]
+  (let [doc-ids (map ensure-document-id ids)
+        docs-with-indices (get-docs-with-indices conn-state doc-ids {})
+        {authorized true forbidden-write false}
+        (group-by #(allow-write? (:_source %) ident)
+                  docs-with-indices)
+        {forbidden true not-visible false}
+        (group-by #(allow-read? (:_source %) ident get-in-config)
+                  forbidden-write)
+        missing (set/difference (set doc-ids)
+                                (set (map :_id docs-with-indices)))
+        not-found (into (map :_id not-visible) missing)
+        prepared-docs (map #(select-keys % [:_index :_type :_id])
+                           authorized)]
+    (cond-> {}
+      forbidden (assoc-in [:errors :forbidden] (map :_id forbidden))
+      (seq not-found) (assoc-in [:errors :not-found] not-found)
+      authorized (assoc :prepared prepared-docs))))
+
+(s/defn bulk-delete :- BulkResult
+  [{:keys [conn] :as conn-state}
+   ids :- [s/Str]
+   ident
+   es-params]
+  (let [{:keys [prepared errors]} (check-and-prepare-bulk conn-state ids ident)
+        bulk-res (when prepared
+                   (try
+                     (format-bulk-res
+                      (ductile.doc/bulk-delete-docs conn
+                                                    prepared
+                                                    (prepare-opts conn-state es-params)))
+                     (catch Exception e
+                       (log/error (str "bulk delete failed: " (.getMessage e))
+                                  (pr-str prepared))
+                       {:errors {:internal-error (map :_id prepared)}})))]
+    (cond-> bulk-res
+      errors (update :errors
+                     #(merge-with concat errors %)))))
+
+(s/defn bulk-update
+  "Generate an ES bulk update handler using some mapping and schema"
+  [Model]
+  (s/fn :- BulkResult
+    [{:keys [conn] :as conn-state}
+     docs :- [Model]
+     ident
+     es-params]
+    (let [by-id (group-by :id docs)
+          ids (seq (keys by-id))
+          {:keys [prepared errors]} (check-and-prepare-bulk conn-state
+                                                            ids
+                                                            ident)
+
+          prepared-docs (map (fn [meta]
+                               (-> (:_id meta)
+                                   by-id
+                                   first
+                                   (into meta)))
+                             prepared)
```

Comment:
While this looks easier to read for me please check this is not a lot slower than `group-by`.
Ideally I would probably had done something along `(let [by-id (fmap first (group-by :id docs)),,,`:

```suggestion
    (let [by-id (into {} (for [{:keys [id] :as doc} docs] [id doc])
          ids (seq (keys by-id))
          {:keys [prepared errors]} (check-and-prepare-bulk conn-state
                                                            ids
                                                            ident)

          prepared-docs (map (fn [{:keys [_id] :as meta}]
                               (into (get by-id _id) meta))
                             prepared)
```

Reply:
group-by seems faster
``` clojure
user> (def docs (map #(array-map :id % :title (str %)) (range 1000)))
;; => #'user/docs
user> (dotimes [n 10]
         (println "(group-by :id docs)")
         (time (group-by :id docs))
         (println "(into {} (for [{:keys [id] :as doc} docs] [id doc]))")
         (time (into {} (for [{:keys [id] :as doc} docs] [id doc]))))
(group-by :id docs)
"Elapsed time: 1.010702 msecs"
(into {} (for [{:keys [id] :as doc} docs] [id doc]))
"Elapsed time: 1.520924 msecs"
(group-by :id docs)
"Elapsed time: 0.361242 msecs"
(into {} (for [{:keys [id] :as doc} docs] [id doc]))
"Elapsed time: 0.516849 msecs"
(group-by :id docs)
"Elapsed time: 0.368968 msecs"
(into {} (for [{:keys [id] :as doc} docs] [id doc]))
"Elapsed time: 0.591694 msecs"
(group-by :id docs)
"Elapsed time: 0.523959 msecs"
(into {} (for [{:keys [id] :as doc} docs] [id doc]))
"Elapsed time: 0.627433 msecs"
(group-by :id docs)
"Elapsed time: 0.384577 msecs"
(into {} (for [{:keys [id] :as doc} docs] [id doc]))
"Elapsed time: 0.735438 msecs"
(group-by :id docs)
"Elapsed time: 0.359023 msecs"
(into {} (for [{:keys [id] :as doc} docs] [id doc]))
"Elapsed time: 0.400209 msecs"
(group-by :id docs)
"Elapsed time: 0.361053 msecs"
(into {} (for [{:keys [id] :as doc} docs] [id doc]))
"Elapsed time: 0.400394 msecs"
(group-by :id docs)
"Elapsed time: 0.353588 msecs"
(into {} (for [{:keys [id] :as doc} docs] [id doc]))
"Elapsed time: 0.420125 msecs"
(group-by :id docs)
"Elapsed time: 0.354192 msecs"
(into {} (for [{:keys [id] :as doc} docs] [id doc]))
"Elapsed time: 0.390267 msecs"
(group-by :id docs)
"Elapsed time: 0.366972 msecs"
(into {} (for [{:keys [id] :as doc} docs] [id doc]))
"Elapsed time: 0.390375 msecs"
```